File: ios/SsdidWallet/Domain/Revocation/RevocationManager.swift
import Foundation

protocol StatusListFetcher: Sendable {
    func fetch(url: String) async throws -> StatusListCredential
}

actor RevocationManager {
    private let fetcher: StatusListFetcher
    private var cache: [String: StatusListCredential] = [:]

    init(fetcher: StatusListFetcher) {
        self.fetcher = fetcher
    }

    func checkRevocation(_ credential: VerifiableCredential) async -> RevocationStatus {
        guard let status = credential.credentialStatus else {
            return .valid
        }

        let listUrl = status.statusListCredential
        guard let index = Int(status.statusListIndex) else {
            return .unknown
        }

        let statusListCred: StatusListCredential
        if let cached = cache[listUrl] {
            statusListCred = cached
        } else {
            do {
                let fetched = try await fetcher.fetch(url: listUrl)
                cache[listUrl] = fetched
                statusListCred = fetched
            } catch {
                return .unknown
            }
        }

        do {
            if try BitstringParser.isRevoked(encodedList: statusListCred.credentialSubject.encodedList, index: index) {
                return .revoked
            } else {
                return .valid
            }
        } catch {
            return .unknown
        }
    }

    func invalidateCache() {
        cache.removeAll()
    }
}

File: android/app/src/main/java/my/ssdid/wallet/domain/revocation/RevocationManager.kt
package my.ssdid.wallet.domain.revocation

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import my.ssdid.wallet.domain.model.VerifiableCredential
import java.util.concurrent.ConcurrentHashMap

enum class RevocationStatus { VALID, REVOKED, UNKNOWN }

fun interface StatusListFetcher {
    suspend fun fetch(url: String): Result<StatusListCredential>
}

class RevocationManager(private val fetcher: StatusListFetcher) {

    private val cache = ConcurrentHashMap<String, StatusListCredential>()
    private val fetchMutex = Mutex()

    suspend fun checkRevocation(credential: VerifiableCredential): RevocationStatus {
        val status = credential.credentialStatus ?: return RevocationStatus.VALID

        val listUrl = status.statusListCredential
        val index = status.statusListIndex.toIntOrNull()
            ?: return RevocationStatus.UNKNOWN

        val statusListCred = cache[listUrl] ?: fetchMutex.withLock {
            cache[listUrl] ?: run {
                val result = fetcher.fetch(listUrl)
                if (result.isFailure) return RevocationStatus.UNKNOWN
                result.getOrThrow().also { cache[listUrl] = it }
            }
        }

        return try {
            if (BitstringParser.isRevoked(statusListCred.credentialSubject.encodedList, index))
                RevocationStatus.REVOKED
            else
                RevocationStatus.VALID
        } catch (_: Exception) {
            RevocationStatus.UNKNOWN
        }
    }

    fun invalidateCache() {
        cache.clear()
    }
}
